src/utils/hooks/useGetResponsesAndElementState.tsx

```diff
@@ -0,0 +1,127 @@
+
+const useGetResponsesAndElementState = (props: { serialNumber: string }) => {
+  const { serialNumber } = props
+  const [responsesByCode, setResponsesByCode] = useState({})
+  const [responsesFullByCode, setResponsesFullByCode] = useState({})
+  const [elementsExpressions, setElementsExpressions] = useState([{}])
+  const [elementsState, setElementsState] = useState({})
+  const [error, setError] = useState('')
+  const [loading, setLoading] = useState(true)
+  const { data, loading: apolloLoading, error: apolloError } = useGetElementsAndResponsesQuery({
+    variables: { serial: serialNumber },
+  })
+
+  useEffect(() => {
+    const error = checkForApplicationErrors(data)
+
+    if (error) {
+      setError(error)
+      setLoading(false)
+      return
+    }
+
+    if (apolloError) return
+
+    const applicationResponses = data?.applicationBySerial?.applicationResponses?.nodes
+
+    const templateSections = data?.applicationBySerial?.template?.templateSections?.nodes
+
+    const templateElements: TemplateElementState[] = []
+
+    templateSections?.forEach((sectionNode) => {
+      sectionNode?.templateElementsBySectionId?.nodes.forEach((element) => {
+        if (element) templateElements.push(element as TemplateElementState) // No idea why ...[spread] doesn't work here.
+      })
+    })
+
+    const currentResponses = {} as ResponsesByCode
+    const currentFullResponses = {} as ResponsesFullByCode
+
+    if (applicationResponses) {
+      applicationResponses.forEach((response) => {
+        const code = response?.templateElement?.code
+        if (code) {
+          currentResponses[code] = response?.value?.text
+          currentFullResponses[code] = response?.value
+        }
+      })
+    }
+
+    setResponsesByCode(currentResponses)
+    setResponsesFullByCode(currentFullResponses)
+    setElementsExpressions(templateElements)
+    setLoading(false)
+  }, [data, apolloError])
+
+  useEffect(() => {
+    evaluateElementExpressions().then((result) => setElementsState(result))
+  }, [responsesByCode])
+
+  async function evaluateElementExpressions() {
+    const promiseArray: Promise<EvaluatedElement>[] = []
+    elementsExpressions.forEach((element) => {
+      promiseArray.push(evaluateSingleElement(element))
+    })
+    const evaluatedElements = await Promise.all(promiseArray)
+    const elementsState: any = {}
+    evaluatedElements.forEach((element) => {
+      elementsState[element.code] = {
+        id: element.id,
+        category: element.category,
+        isEditable: element.isEditable,
+        isRequired: element.isRequired,
+        isVisible: element.isVisible,
+        // isValid: element.isValid,
+      }
+    })
+    return elementsState
+  }
+
+  async function evaluateSingleElement(element: any): Promise<EvaluatedElement> {
+    const evaluationParameters = {
+      objects: [responsesByCode, responsesFullByCode], // TO-DO: Also send user/org objects etc.
+      // graphQLConnection: TO-DO
+    }
```

Comment:
This is not in use?

Reply:
What do you mean? What is not in use?

Do you mean the graphQLConnection? No, there is no expressions that use it yet, but we can pass this object in here quite easily when we need it.

Reply:
Sorry, the `evaluationParameters`.

Reply:
Ah, right, yes -- good catch, they should be passed in to the `evaluateExpression` function.